### ISSUE TITLE
[MNG-723] [patch] assembly:unpack

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#3772]
**[Jason Grant](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=logularjason)** opened **[MNG-723](https://issues.apache.org/jira/browse/MNG-723?redirect=false)** and commented

I've coded an assembly:unpack goal.  It will unpack any dependencies in the project whose filenames indicate type jar or zip.  These are unpacked into the workDirectory.

I went about it by pulling common infrastructure out of AssemblyMojo and placing it in a new abstract superclass AbstractAssemblyMojo.  Another new class, called UnpackMojo now also resides in maven-assembly-plugin.

I was going to code an integration test for this, however I will instead do some work on http://jira.codehaus.org/browse/MNG-483 first, so that I can use the latter as the first step in the test.  I'll write some supporting notes over on that jira issue.

Note that this is the first code that I've contributed to Maven - I'm happy to take direction should I be breaking conventions.  You'll see me from time to time on #maven as jasong.

---

**Affects:** 2.0-beta-1

**Attachments:**
- [AbstractAssemblyMojo.java](https://issues.apache.org/jira/secure/attachment/12711573/AbstractAssemblyMojo.java) (_9.52 kB_)
- [assembly-050811-1.zip](https://issues.apache.org/jira/secure/attachment/12711784/assembly-050811-1.zip) (_6.11 kB_)
- [AssemblyMojo.diff](https://issues.apache.org/jira/secure/attachment/12711644/AssemblyMojo.diff) (_21.35 kB_)
- [UnpackMojo.java](https://issues.apache.org/jira/secure/attachment/12711505/UnpackMojo.java) (_3.40 kB_)
- [UnpackMojo.java](https://issues.apache.org/jira/secure/attachment/12711646/UnpackMojo.java) (_3.61 kB_)

**Comment by jira-importer** [comment]
**[Jason Grant](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=logularjason)** commented

The first attachment still had System.out debug statements.  This one doesnt.

**Comment by jira-importer** [comment]
**[Jason Grant](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=logularjason)** commented

I want to do a further update - AssemblyMojo has been wired so far to assume that dependencies are jars - if unpack is required, then it tries to unjar the dependency.  I'd like to contribute an additional enhancement to allow it to handle zips too.

Also, reading my code just now, I think more methods should be pulled down from the abstract class.  I won't burden you with more diffs just now, but I might ask if it is OK to check this stuff in rather than polluting jira issues with superceded versions (I cannot delete my attachments above).

**Comment by jira-importer** [comment]
**[Jason Grant](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=logularjason)** commented

Oh what the heck - did the changes described above and am attaching the code, since doing otherwise will just hold things up.  This time, I'm zipping the source and attaching it as assembly-050811-1.zip.  The other attachments are now obsolete.

**Comment by jira-importer** [comment]
**[Jason Grant](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=logularjason)** commented

The code attached to this issue has now been enhanced with extra functionality as described in #MNG-483.

I've attached a new zip to that issue; it now obsoletes all of the attachments here.  Actually, the only change is a new method in AssemblyMojo, but I've attached the same 3 classes there to make it easy to push into svn.
